<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SPatacseq • SPatacseq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="SPatacseq">
<meta property="og:description" content="SPatacseq">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-CCMXLYXWYV"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-CCMXLYXWYV');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPatacseq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/SPatacseq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/systemPipeR/SPatacseq/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="SPatacseq_files/header-attrs-2.7/header-attrs.js"></script><script src="SPatacseq_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="SPatacseq_files/jquery-3.5.1/jquery.min.js"></script><link href="SPatacseq_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="SPatacseq_files/datatables-binding-0.18/datatables.js"></script><link href="SPatacseq_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="SPatacseq_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="SPatacseq_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="SPatacseq_files/dt-ext-fixedcolumns-1.10.20/css/fixedColumns.dataTables.min.css" rel="stylesheet">
<script src="SPatacseq_files/dt-ext-fixedcolumns-1.10.20/js/dataTables.fixedColumns.min.js"></script><link href="SPatacseq_files/dt-ext-scroller-1.10.20/css/scroller.dataTables.min.css" rel="stylesheet">
<script src="SPatacseq_files/dt-ext-scroller-1.10.20/js/dataTables.scroller.min.js"></script><link href="SPatacseq_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="SPatacseq_files/crosstalk-1.1.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SPatacseq</h1>
                        <h4 class="author">Author: Daniela Cassol (<a href="mailto:danielac@ucr.edu" class="email">danielac@ucr.edu</a>) and Thomas Girke (<a href="mailto:thomas.girke@ucr.edu" class="email">thomas.girke@ucr.edu</a>)</h4>
            
            <h4 class="date">Last update: 23 April, 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/systemPipeR/SPatacseq/blob/master/vignettes/SPatacseq.Rmd"><code>vignettes/SPatacseq.Rmd</code></a></small>
      <div class="hidden name"><code>SPatacseq.Rmd</code></div>

    </div>

    
    
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<!--
- Compile from command-line
Rscript -e "rmarkdown::render('SPatacseq.Rmd', c('BiocStyle::html_document'), clean=TRUE)"
--><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<blockquote>
<p>Users want to provide here background information about the design of their ATAC-Seq project.</p>
</blockquote>
<p><strong>ATAC-seq</strong> (<strong>A</strong>ssay for <strong>T</strong>ransposase-<strong>A</strong>ccessible <strong>C</strong>hromatin with high-throughput <strong>seq</strong>uencing) is a method to investigate the chromatin accessibility at a genome-wide level.</p>
<div class="figure" style="text-align: center">
<img src="Buenrostro2013.png" alt="ATAC-seq schematic methodology to chromatin accessibility using the Tn5 transposase.
         In the open chromatin (line between nucleosomes in gray), Tn5 transposase (green) cuts          the regions and insert the adaptors (red and blue) to generate sequencing-library
        fragments. [@Buenrostro2013-ap]" width="1821"><p class="caption">
ATAC-seq schematic methodology to chromatin accessibility using the Tn5 transposase. In the open chromatin (line between nucleosomes in gray), Tn5 transposase (green) cuts the regions and insert the adaptors (red and blue) to generate sequencing-library fragments. <span class="citation">(Buenrostro et al. 2013)</span>
</p>
</div>
<p><code>SPatacseq</code> pipeline will provide the steps to perform an ATAC-seq analysis. The pipeline show alternatives to trim adapters, check raw sequence read quality, alignment to the reference genome, call peaks, and visualization report.</p>
</div>
<div id="samples-and-environment-settings" class="section level1">
<h1 class="hasAnchor">
<a href="#samples-and-environment-settings" class="anchor"></a>Samples and environment settings</h1>
<div id="required-packages-and-resources" class="section level2">
<h2 class="hasAnchor">
<a href="#required-packages-and-resources" class="anchor"></a>Required packages and resources</h2>
<p>To go through this tutorial, you need the following software installed:</p>
<ul>
<li>R/&gt;=4.0.3</li>
<li>systemPipeR R package (version 1.24)</li>
<li>bwa/0.7.17</li>
</ul>
<p>The <code>systemPipeR</code> package <span class="citation">(H Backman and Girke 2016)</span> needs to be loaded to perform the analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://systempipe.org/">systemPipeR</a></span><span class="op">)</span></code></pre></div>
<p>If you desire to build your pipeline with any different software, make sure to have the respective software installed and configured in your <code>PATH.</code> To make sure if the configuration is right, you always can test as follow:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/tryCL.html">tryCL</a></span><span class="op">(</span>command <span class="op">=</span> <span class="st">"bwa"</span><span class="op">)</span></code></pre></div>
</div>
<div id="structure-of-targets-file" class="section level2">
<h2 class="hasAnchor">
<a href="#structure-of-targets-file" class="anchor"></a>Structure of <em><code>targets</code></em> file</h2>
<p>The <em><code>targets</code></em> file defines all input files (<em>e.g.</em> FASTQ, BAM, BCF) and sample comparisons of an analysis workflow. The following shows the format of a sample <em><code>targets</code></em> file included in the package. It also can be viewed and downloaded from <em><code>systemPipeR</code></em>’s GitHub repository <a href="https://github.com/tgirke/systemPipeR/blob/master/inst/extdata/targets.txt">here</a>. In a target file with a single type of input files, here FASTQ files of single-end (SE) reads, the first three columns are mandatory including their column names, while it is four mandatory columns for FASTQ files of PE reads. All subsequent columns are optional and any number of additional columns can be added as needed.</p>
<p>Users should note here, the usage of targets files is optional when using <em>systemPipeR’s</em> new CWL interface. They can be replaced by a standard YAML input file used by CWL. Since for organizing experimental variables targets files are extremely useful and user-friendly. Thus, we encourage users to keep using them.</p>
<div id="structure-of-targets-file-for-single-end-se-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#structure-of-targets-file-for-single-end-se-samples" class="anchor"></a>Structure of <em><code>targets</code></em> file for single-end (SE) samples</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">targetspath</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="op">]</span></code></pre></div>
<pre><code>##                      FileName SampleName Factor SampleLong
## 1 ./data/SRR446027_1.fastq.gz        M1A     M1  Mock.1h.A
## 2 ./data/SRR446028_1.fastq.gz        M1B     M1  Mock.1h.B
## 3 ./data/SRR446029_1.fastq.gz        A1A     A1   Avr.1h.A
## 4 ./data/SRR446030_1.fastq.gz        A1B     A1   Avr.1h.B
##   Experiment        Date
## 1          1 23-Mar-2012
## 2          1 23-Mar-2012
## 3          1 23-Mar-2012
## 4          1 23-Mar-2012</code></pre>
<p>To work with custom data, users need to generate a <em><code>targets</code></em> file containing the paths to their own FASTQ files and then provide under <em><code>targetspath</code></em> the path to the corresponding <em><code>targets</code></em> file.</p>
</div>
</div>
</div>
<div id="quality-control" class="section level1">
<h1 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality control</h1>
<div id="read-preprocessing" class="section level2">
<h2 class="hasAnchor">
<a href="#read-preprocessing" class="anchor"></a>Read Preprocessing</h2>
<div id="preprocessing-with-preprocessreads-function" class="section level3">
<h3 class="hasAnchor">
<a href="#preprocessing-with-preprocessreads-function" class="anchor"></a>Preprocessing with <em><code>preprocessReads</code></em> function</h3>
<p>The function <em><code>preprocessReads</code></em> allows to apply predefined or custom read preprocessing functions to all FASTQ files referenced in a <em><code>SYSargs2</code></em> container, such as quality filtering or adaptor trimming routines. The paths to the resulting output FASTQ files are stored in the <em><code>output</code></em> slot of the <em><code>SYSargs2</code></em> object. Internally, <em><code>preprocessReads</code></em> uses the <em><code>FastqStreamer</code></em> function from the <em><code>ShortRead</code></em> package to stream through large FASTQ files in a memory-efficient manner. The following example performs adaptor trimming with the <em><code>trimLRPatterns</code></em> function from the <em><code>Biostrings</code></em> package. After the trimming step a new targets file is generated (here <em><code>targets_trimPE.txt</code></em>) containing the paths to the trimmed FASTQ files. The new targets file can be used for the next workflow step with an updated <em><code>SYSargs2</code></em> instance, <em>e.g.</em> running the NGS alignments with the trimmed FASTQ files.</p>
<p>Construct <em><code>SYSargs2</code></em> object from <em><code>cwl</code></em> and <em><code>yml</code></em> param and <em><code>targets</code></em> files.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetsPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/preprocessReads/trim-pe"</span>, 
    package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">trim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetsPE</span>, wf_file <span class="op">=</span> <span class="st">"trim-pe.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"trim-pe.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">trim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">trim</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">trim</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">output</a></span><span class="op">(</span><span class="va">trim</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/preprocessReads.html">preprocessReads</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">trim</span>, Fct <span class="op">=</span> <span class="st">"trimLRPatterns(Rpattern='GCCCGGGTAA', 
                subject=fq)"</span>, 
    batchsize <span class="op">=</span> <span class="fl">1e+05</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span>, compress <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>The following example shows how one can design a custom read preprocessing function using utilities provided by the <em><code>ShortRead</code></em> package, and then run it in batch mode with the <em>‘preprocessReads’</em> function (here on paired-end reads).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">filterFct</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">fq</span>, <span class="va">cutoff</span> <span class="op">=</span> <span class="fl">20</span>, <span class="va">Nexceptions</span> <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">qcount</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu">as</span><span class="op">(</span><span class="fu">quality</span><span class="op">(</span><span class="va">fq</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="va">cutoff</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
    <span class="co"># Retains reads where Phred scores are &gt;= cutoff with N</span>
    <span class="co"># exceptions</span>
    <span class="va">fq</span><span class="op">[</span><span class="va">qcount</span> <span class="op">&lt;=</span> <span class="va">Nexceptions</span><span class="op">]</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/preprocessReads.html">preprocessReads</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">trim</span>, Fct <span class="op">=</span> <span class="st">"filterFct(fq, cutoff=20, Nexceptions=0)"</span>, 
    batchsize <span class="op">=</span> <span class="fl">1e+05</span><span class="op">)</span></code></pre></div>
</div>
<div id="preprocessing-with-trimgalore" class="section level3">
<h3 class="hasAnchor">
<a href="#preprocessing-with-trimgalore" class="anchor"></a>Preprocessing with TrimGalore!</h3>
<p><a href="http://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">TrimGalore!</a> is a wrapper tool to consistently apply quality and adapter trimming to fastq files, with some extra functionality for removing Reduced Representation Bisulfite-Seq (RRBS) libraries.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targets.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/trim_galore/trim_galore-se"</span>, 
    package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">trimG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targets</span>, wf_file <span class="op">=</span> <span class="st">"trim_galore-se.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"trim_galore-se.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">trimG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">trimG</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">trimG</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">cmdlist</a></span><span class="op">(</span><span class="va">trimG</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">output</a></span><span class="op">(</span><span class="va">trimG</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="co">## Run Single Machine Option</span>
<span class="va">trimG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">trimG</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
<div id="preprocessing-with-trimmomatic" class="section level3">
<h3 class="hasAnchor">
<a href="#preprocessing-with-trimmomatic" class="anchor"></a>Preprocessing with Trimmomatic</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetsPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/trimmomatic/trimmomatic-pe"</span>, 
    package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">trimM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetsPE</span>, wf_file <span class="op">=</span> <span class="st">"trimmomatic-pe.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"trimmomatic-pe.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">trimM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">trimM</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">trimM</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">cmdlist</a></span><span class="op">(</span><span class="va">trimM</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">output</a></span><span class="op">(</span><span class="va">trimM</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="co">## Run Single Machine Option</span>
<span class="va">trimM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">trimM</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="fastq-quality-report" class="section level2">
<h2 class="hasAnchor">
<a href="#fastq-quality-report" class="anchor"></a>FASTQ quality report</h2>
<p>The following <em><code>seeFastq</code></em> and <em><code>seeFastqPlot</code></em> functions generate and plot a series of useful quality statistics for a set of FASTQ files including per cycle quality box plots, base proportions, base-level quality trends, relative k-mer diversity, length and occurrence distribution of reads, number of reads above quality cutoffs and mean quality distribution.<br>
The function <em><code>seeFastq</code></em> computes the quality statistics and stores the results in a relatively small list object that can be saved to disk with <em><code><a href="https://rdrr.io/r/base/save.html">save()</a></code></em> and reloaded with <em><code><a href="https://rdrr.io/r/base/load.html">load()</a></code></em> for later plotting. The argument <em><code>klength</code></em> specifies the k-mer length and <em><code>batchsize</code></em> the number of reads to a random sample from each FASTQ file.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fqlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html">seeFastq</a></span><span class="op">(</span>fastq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs.html">infile1</a></span><span class="op">(</span><span class="va">trim</span><span class="op">)</span>, batchsize <span class="op">=</span> <span class="fl">10000</span>, 
    klength <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html">pdf</a></span><span class="op">(</span><span class="st">"./results/fastqReport.pdf"</span>, height <span class="op">=</span> <span class="fl">18</span>, width <span class="op">=</span> <span class="fl">4</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">fqlist</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html">seeFastqPlot</a></span><span class="op">(</span><span class="va">fqlist</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<center>
<img src="fastqReport.png">
</center>
<div align="center">
<strong>Figure 5:</strong> FASTQ quality report
</div>
<p></p>
<p>Parallelization of FASTQ quality report on a single machine with multiple cores.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">f</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html">seeFastq</a></span><span class="op">(</span>fastq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs.html">infile1</a></span><span class="op">(</span><span class="va">trim</span><span class="op">)</span><span class="op">[</span><span class="va">x</span><span class="op">]</span>, batchsize <span class="op">=</span> <span class="fl">1e+05</span>, 
    klength <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="va">fqlist</span> <span class="op">&lt;-</span> <span class="fu">bplapply</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">trim</span><span class="op">)</span>, <span class="va">f</span>, BPPARAM <span class="op">=</span> <span class="fu">MulticoreParam</span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html">seeFastqPlot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">fqlist</span>, recursive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Parallelization of FASTQ quality report via scheduler (<em>e.g.</em> Slurm) across several compute nodes.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mllg/batchtools">batchtools</a></span><span class="op">)</span>
<span class="va">f</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://systempipe.org/">systemPipeR</a></span><span class="op">)</span>
    <span class="va">targetsPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
    <span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/preprocessReads/trim-pe"</span>, 
        package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
    <span class="va">trim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetsPE</span>, wf_file <span class="op">=</span> <span class="st">"trim-pe.cwl"</span>, 
        input_file <span class="op">=</span> <span class="st">"trim-pe.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
    <span class="va">trim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">trim</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
        FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html">seeFastq</a></span><span class="op">(</span>fastq <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs.html">infile1</a></span><span class="op">(</span><span class="va">trim</span><span class="op">)</span><span class="op">[</span><span class="va">x</span><span class="op">]</span>, batchsize <span class="op">=</span> <span class="fl">1e+05</span>, klength <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BatchtoolsParam-class.html">BatchtoolsParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">4</span>, cluster <span class="op">=</span> <span class="st">"slurm"</span>, template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, 
    resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="va">fqlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">bplapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">trim</span><span class="op">)</span>, <span class="va">f</span>, BPPARAM <span class="op">=</span> <span class="va">param</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html">seeFastqPlot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">fqlist</span>, recursive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="read-alignment" class="section level1">
<h1 class="hasAnchor">
<a href="#read-alignment" class="anchor"></a>Read Alignment</h1>
<p>After quality control, the sequence reads can be aligned to a reference genome or miRNA database. The following sessions present some NGS sequence alignment software. Select the most accurate aligner and determining the optimal parameter for your custom data set project.</p>
<p>For all the following examples, it is necessary to install the respective software and export the PATH accordingly. If it is available Environment Module in the system, you can load all the request software with <code><a href="https://rdrr.io/pkg/systemPipeR/man/moduleload.html">moduleload(modules(idx))</a></code> function.</p>
<div id="alignment-with-bowtie2" class="section level2">
<h2 class="hasAnchor">
<a href="#alignment-with-bowtie2" class="anchor"></a>Alignment with <em><code>Bowtie2</code></em>
</h2>
<p>The following example runs <em><code>Bowtie2</code></em> as a single process without submitting it to a cluster.</p>
<p>Building the index:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/bowtie2/bowtie2-idx"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"bowtie2-index.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"bowtie2-index.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>
<span class="va">idx</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">cmdlist</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>

<span class="co">## Run in single machine</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">idx</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>Building all the command-line:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/bowtie2/bowtie2-mi"</span>, package <span class="op">=</span> <span class="st">"SPatacseq"</span><span class="op">)</span>
<span class="va">bowtie</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetspath</span>, wf_file <span class="op">=</span> <span class="st">"bowtie2-mapping-mi.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"bowtie2-mapping-mi.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">bowtie</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">bowtie</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileTrim <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">bowtie</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">cmdlist</a></span><span class="op">(</span><span class="va">bowtie</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">output</a></span><span class="op">(</span><span class="va">bowtie</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></code></pre></div>
<p>Please note that each experiment and/or each species may require an optimization of the parameters used. Here is an example where no mismatches are allowed, and <code>-k</code> mode is used. The <code>-k</code> represents the maximum number of alignments to report per read. By default, Bowtie2 reports at most one alignment per read, and if multiple equivalent alignments exist, it chooses one randomly. There is indicative that <code>Bowtie2</code> with very sensitive local (<code>--very-sensitive-local</code>) argument provides better accuracy and generates smaller <code>P-values</code> for true positives <span class="citation">(Ziemann, Kaspi, and El-Osta 2016)</span>.</p>
<p>Running all the jobs to computing nodes.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">bowtie</span>, FUN <span class="op">=</span> <span class="va">runCommandline</span>, more.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">bowtiePE</span>, 
    dir <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, conffile <span class="op">=</span> <span class="st">".batchtools.conf.R"</span>, template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, 
    Njobs <span class="op">=</span> <span class="fl">6</span>, runid <span class="op">=</span> <span class="st">"01"</span>, resourceList <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/getStatus.html">getStatus</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span>
<span class="va">bowtie</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/output_update.html">output_update</a></span><span class="op">(</span><span class="va">bowtie</span>, dir <span class="op">=</span> <span class="cn">FALSE</span>, replace <span class="op">=</span> <span class="cn">TRUE</span>, 
    extension <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">".sam"</span>, <span class="st">".bam"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Alternatively, it possible to run all the jobs in a single machine.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bowtie</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">bowtie</span>, make_bam <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">systemPipeR</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/check.output.html">check.output</a></span><span class="op">(</span><span class="va">bowtie</span><span class="op">)</span></code></pre></div>
<p>Create new targets file.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">clt</a></span><span class="op">(</span><span class="va">bowtie</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">bowtie</span>, file <span class="op">=</span> <span class="st">"default"</span>, step <span class="op">=</span> <span class="fl">1</span>, new_col <span class="op">=</span> <span class="st">"bowtie"</span>, 
    new_col_output_index <span class="op">=</span> <span class="fl">1</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div id="read-and-alignment-count-stats" class="section level3">
<h3 class="hasAnchor">
<a href="#read-and-alignment-count-stats" class="anchor"></a>Read and alignment count stats</h3>
<p>Generate a table of read and alignment counts for all samples.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">read_statsDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/alignStats.html">alignStats</a></span><span class="op">(</span><span class="va">bowtie</span>, subset <span class="op">=</span> <span class="st">"FileTrim"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">read_statsDF</span>, <span class="st">"results/alignStats_bowtie.xls"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, 
    quote <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></code></pre></div>
<p>The following shows the first four lines of the sample alignment stats file provided by the systemPipeR package.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"alignStats.xls"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="fu">systemPipeR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/showDT.html">showDT</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="va">table</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>  <span class="co">## system</span></code></pre></div>
<div id="htmlwidget-10ba681a8d8ca283aaf7" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-10ba681a8d8ca283aaf7">{"x":{"filter":"none","extensions":["FixedColumns","Scroller"],"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18"],["M1A","M1B","A1A","A1B","V1A","V1B","M6A","M6B","A6A","A6B","V6A","V6B","M12A","M12B","A12A","A12B","V12A","V12B"],[192918,197484,189870,188854,198732,195542,197234,180904,183140,196578,194204,198414,199434,188954,187922,199790,193922,181006],[177961,159378,176055,147768,160115,184418,182469,162636,170258,179932,176857,176825,179722,162894,171326,173771,174180,157790],[92.2469650317752,80.7042595855867,92.7239690314426,78.2445698793777,80.5683030412817,94.3111965715805,92.5139681799284,89.901826383054,92.9660369116523,91.5321144787311,91.0676402133839,89.1192153779471,90.1160283602595,86.2082834975708,91.1686763657262,86.9768256669503,89.819618197007,87.1739058373756],[177961,159378,176055,147768,160115,184418,182469,162636,170258,179932,176857,176825,179722,162894,171326,173771,174180,157790],[92.2469650317752,80.7042595855867,92.7239690314426,78.2445698793777,80.5683030412817,94.3111965715805,92.5139681799284,89.901826383054,92.9660369116523,91.5321144787311,91.0676402133839,89.1192153779471,90.1160283602595,86.2082834975708,91.1686763657262,86.9768256669503,89.819618197007,87.1739058373756]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>FileName<\/th>\n      <th>Nreads2x<\/th>\n      <th>Nalign<\/th>\n      <th>Perc_Aligned<\/th>\n      <th>Nalign_Primary<\/th>\n      <th>Perc_Aligned_Primary<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"fixedColumns":true,"deferRender":true,"scrollY":200,"scroller":true,"columnDefs":[{"className":"dt-right","targets":[2,3,4,5,6]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="alignment-with-bwa-mem-e-g--for-var-seq" class="section level3">
<h3 class="hasAnchor">
<a href="#alignment-with-bwa-mem-e-g--for-var-seq" class="anchor"></a>Alignment with <em><code>BWA-MEM</code></em> (<em>e.g.</em> for VAR-Seq)</h3>
<p>The following example runs BWA-MEM as a single process without submitting it to a cluster. ##TODO: add reference</p>
<p>Build the index:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/bwa/bwa-idx"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"bwa-index.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"bwa-index.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>
<span class="va">idx</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">cmdlist</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>  <span class="co"># Indexes reference genome</span>

<span class="co">## Run</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">idx</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>Running the alignment:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetsPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/bwa/bwa-pe"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">bwaPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWorkflow</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targetsPE</span>, wf_file <span class="op">=</span> <span class="st">"bwa-pe.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"bwa-pe.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">bwaPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">bwaPE</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">bwaPE</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">cmdlist</a></span><span class="op">(</span><span class="va">bwaPE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">output</a></span><span class="op">(</span><span class="va">bwaPE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="co">## Single Machine</span>
<span class="va">bwaPE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runCommandline.html">runCommandline</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">bwaPE</span>, make_bam <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co">## Cluster</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mllg/batchtools">batchtools</a></span><span class="op">)</span>
<span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">120</span>, ntasks <span class="op">=</span> <span class="fl">1</span>, ncpus <span class="op">=</span> <span class="fl">4</span>, memory <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span>
<span class="va">reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/clusterRun.html">clusterRun</a></span><span class="op">(</span><span class="va">bwaPE</span>, FUN <span class="op">=</span> <span class="va">runCommandline</span>, more.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">bwaPE</span>, 
    dir <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, conffile <span class="op">=</span> <span class="st">".batchtools.conf.R"</span>, template <span class="op">=</span> <span class="st">"batchtools.slurm.tmpl"</span>, 
    Njobs <span class="op">=</span> <span class="fl">18</span>, runid <span class="op">=</span> <span class="st">"01"</span>, resourceList <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/batchtools/man/getStatus.html">getStatus</a></span><span class="op">(</span>reg <span class="op">=</span> <span class="va">reg</span><span class="op">)</span></code></pre></div>
<p>Create new targets file.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">clt</a></span><span class="op">(</span><span class="va">bwaPE</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">bwaPE</span>, file <span class="op">=</span> <span class="st">"default"</span>, step <span class="op">=</span> <span class="fl">1</span>, new_col <span class="op">=</span> <span class="st">"bwaPE"</span>, 
    new_col_output_index <span class="op">=</span> <span class="fl">1</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Check whether all BAM files have been created and write out the new targets file.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">bwaPE</span>, file <span class="op">=</span> <span class="st">"targets_bam.txt"</span>, step <span class="op">=</span> <span class="fl">1</span>, 
    new_col <span class="op">=</span> <span class="st">"FileName"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">outpaths</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="create-symbolic-links-for-viewing-bam-files-in-igv" class="section level2">
<h2 class="hasAnchor">
<a href="#create-symbolic-links-for-viewing-bam-files-in-igv" class="anchor"></a>Create symbolic links for viewing BAM files in IGV</h2>
<p>The <code>symLink2bam</code> function creates symbolic links to view the BAM alignment files in a genome browser such as IGV without moving these large files to a local system. The corresponding URLs are written to a file with a path specified under <code>urlfile</code>, here <code>IGVurl.txt</code>. Please replace the directory and the user name.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/symLink2bam.html">symLink2bam</a></span><span class="op">(</span>sysargs <span class="op">=</span> <span class="va">args</span>, htmldir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"~/.html/"</span>, <span class="st">"somedir/"</span><span class="op">)</span>, 
    urlbase <span class="op">=</span> <span class="st">"http://cluster.hpcc.ucr.edu/~tgirke/"</span>, urlfile <span class="op">=</span> <span class="st">"./results/IGVurl.txt"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="utilities-for-coverage-data" class="section level1">
<h1 class="hasAnchor">
<a href="#utilities-for-coverage-data" class="anchor"></a>Utilities for coverage data</h1>
<p>The following introduces several utilities useful for ChIP-Seq data. They are not part of the actual workflow.</p>
<div id="rle-object-stores-coverage-information" class="section level2">
<h2 class="hasAnchor">
<a href="#rle-object-stores-coverage-information" class="anchor"></a>Rle object stores coverage information</h2>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools">Rsamtools</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicAlignments">GenomicAlignments</a></span><span class="op">)</span>
<span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">bwaPE</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">aligns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicAlignments/man/readGAlignments.html">readGAlignments</a></span><span class="op">(</span><span class="va">outpaths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="va">cov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/coverage-methods.html">coverage</a></span><span class="op">(</span><span class="va">aligns</span><span class="op">)</span>
<span class="va">cov</span>
<span class="co"># RleList of length 7 $Chr1 integer-Rle of length 100000 with</span>
<span class="co"># 11827 runs Lengths: 3644 3 2 3 2 1 ...  1 5 23 3 12 Values</span>
<span class="co"># : 0 1 3 4 5 6 ...  17 19 20 19 18 $Chr2 integer-Rle of</span>
<span class="co"># length 100000 with 5085 runs Lengths: 999 7 11 55 2 5 ...</span>
<span class="co"># 2 1 1 1 3357 Values : 0 2 3 4 3 2 ...  8 6 2 1 0 $Chr3</span>
<span class="co"># integer-Rle of length 100000 with 12688 runs Lengths: 8953</span>
<span class="co"># 5 1 2 9 5 ...  3 4 4 1 35 Values : 0 1 2 3 4 5 ...  16 15</span>
<span class="co"># 14 2 0 $Chr4 integer-Rle of length 100000 with 5113 runs</span>
<span class="co"># Lengths: 4104 53 157 71 4 50 ...  43 12 63 12 446 Values :</span>
<span class="co"># 0 1 0 1 2 1 ...  0 1 2 1 0 $Chr5 integer-Rle of length</span>
<span class="co"># 100000 with 6995 runs Lengths: 1236 34 1 2 1 1 ...  13 6 1</span>
<span class="co"># 55 413 Values : 0 1 2 3 4 5 ...  5 4 3 1 0 ...  &lt;2 more</span>
<span class="co"># elements&gt;</span></code></pre></div>
</div>
<div id="resizing-aligned-reads" class="section level2">
<h2 class="hasAnchor">
<a href="#resizing-aligned-reads" class="anchor"></a>Resizing aligned reads</h2>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/intra-range-methods.html">trim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/intra-range-methods.html">resize</a></span><span class="op">(</span><span class="fu">as</span><span class="op">(</span><span class="va">aligns</span>, <span class="st">"GRanges"</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">200</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="naive-peak-calling" class="section level2">
<h2 class="hasAnchor">
<a href="#naive-peak-calling" class="anchor"></a>Naive peak calling</h2>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">islands</span> <span class="op">&lt;-</span> <span class="fu">slice</span><span class="op">(</span><span class="va">cov</span>, lower <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>
<span class="va">islands</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
</div>
<div id="plot-coverage-for-defined-region" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-coverage-for-defined-region" class="anchor"></a>Plot coverage for defined region</h2>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tengfei.github.com/ggbio/">ggbio</a></span><span class="op">)</span>
<span class="va">myloc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Chr1"</span>, <span class="fl">1</span>, <span class="fl">1e+05</span><span class="op">)</span>
<span class="va">ga</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicAlignments/man/readGAlignments.html">readGAlignments</a></span><span class="op">(</span><span class="va">outpaths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, use.names <span class="op">=</span> <span class="cn">TRUE</span>, param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html">ScanBamParam</a></span><span class="op">(</span>which <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="va">myloc</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, 
    <span class="fu">IRanges</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">myloc</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">myloc</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ggbio/man/autoplot-method.html">autoplot</a></span><span class="op">(</span><span class="va">ga</span>, <span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">strand</span>, fill <span class="op">=</span> <span class="va">strand</span><span class="op">)</span>, facets <span class="op">=</span> <span class="va">strand</span> <span class="op">~</span> 
    <span class="va">seqnames</span>, stat <span class="op">=</span> <span class="st">"coverage"</span><span class="op">)</span></code></pre></div>
<p><img src="ga.png"></p>
</div>
</div>
<div id="peak-calling-with-macs2" class="section level1">
<h1 class="hasAnchor">
<a href="#peak-calling-with-macs2" class="anchor"></a>Peak calling with MACS2</h1>
<div id="merge-bam-files-of-replicates-prior-to-peak-calling" class="section level2">
<h2 class="hasAnchor">
<a href="#merge-bam-files-of-replicates-prior-to-peak-calling" class="anchor"></a>Merge BAM files of replicates prior to peak calling</h2>
<p>Merging BAM files of technical and/or biological replicates can improve the sensitivity of the peak calling by increasing the depth of read coverage. The <code>mergeBamByFactor</code> function merges BAM files based on grouping information specified by a <code>factor</code>, here the <code>Factor</code> column of the imported targets file. It also returns an updated <code>SYSargs2</code> object containing the paths to the merged BAM files as well as to any unmerged files without replicates. This step can be skipped if merging of BAM files is not desired.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/mergeBamByFactor"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWF</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="st">"targets_bam.txt"</span>, wf_file <span class="op">=</span> <span class="st">"merge-bam.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"merge-bam.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_BAM_PATH_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>

<span class="va">args_merge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/mergeBamByFactor.html">mergeBamByFactor</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">args</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">args_merge</span>, file <span class="op">=</span> <span class="st">"targets_mergeBamByFactor.txt"</span>, 
    step <span class="op">=</span> <span class="fl">1</span>, new_col <span class="op">=</span> <span class="st">"FileName"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Skip if a module system is not used</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/moduleload.html">module</a></span><span class="op">(</span><span class="st">"list"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/moduleload.html">module</a></span><span class="op">(</span><span class="st">"unload"</span>, <span class="st">"miniconda2"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/moduleload.html">module</a></span><span class="op">(</span><span class="st">"load"</span>, <span class="st">"python/2.7.14"</span><span class="op">)</span>  <span class="co"># Make sure to set up your enviroment variable for MACS2 </span></code></pre></div>
</div>
<div id="peak-calling-without-inputreference-sample" class="section level2">
<h2 class="hasAnchor">
<a href="#peak-calling-without-inputreference-sample" class="anchor"></a>Peak calling without input/reference sample</h2>
<p>MACS2 can perform peak calling on ChIP-Seq data with and without input samples <span class="citation">(Zhang et al. 2008)</span>. The following performs peak calling without input on all samples specified in the corresponding <code>args</code> object. Note, due to the small size of the sample data, MACS2 needs to be run here with the <code>nomodel</code> setting. For real data sets, users want to remove this parameter in the corresponding <code>*.param</code> file(s).</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/MACS2/MACS2-noinput/"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWF</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="st">"targets_mergeBamByFactor.txt"</span>, wf_file <span class="op">=</span> <span class="st">"macs2.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"macs2.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">args</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span>, force <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">outpaths</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">args</span>, file <span class="op">=</span> <span class="st">"targets_macs.txt"</span>, step <span class="op">=</span> <span class="fl">1</span>, 
    new_col <span class="op">=</span> <span class="st">"FileName"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="peak-calling-with-inputreference-sample" class="section level2">
<h2 class="hasAnchor">
<a href="#peak-calling-with-inputreference-sample" class="anchor"></a>Peak calling with input/reference sample</h2>
<p>To perform peak calling with input samples, they can be most conveniently specified in the <code>SampleReference</code> column of the initial <code>targets</code> file. The <code>writeTargetsRef</code> function uses this information to create a <code>targets</code> file intermediate for running MACS2 with the corresponding input samples.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsRef.html">writeTargetsRef</a></span><span class="op">(</span>infile <span class="op">=</span> <span class="st">"targets_mergeBamByFactor.txt"</span>, outfile <span class="op">=</span> <span class="st">"targets_bam_ref.txt"</span>, 
    silent <span class="op">=</span> <span class="cn">FALSE</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/MACS2/MACS2-input/"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">args_input</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWF</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="st">"targets_bam_ref.txt"</span>, wf_file <span class="op">=</span> <span class="st">"macs2-input.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"macs2.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">args_input</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args_input</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs2-methods.html">cmdlist</a></span><span class="op">(</span><span class="va">args_input</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>
<span class="co">## Run</span>
<span class="va">args_input</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runCommandline.html">runCommandline</a></span><span class="op">(</span><span class="va">args_input</span>, make_bam <span class="op">=</span> <span class="cn">FALSE</span>, force <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="va">outpaths_input</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args_input</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, 
    index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">outpaths_input</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">args_input</span>, file <span class="op">=</span> <span class="st">"targets_macs_input.txt"</span>, 
    step <span class="op">=</span> <span class="fl">1</span>, new_col <span class="op">=</span> <span class="st">"FileName"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>The peak calling results from MACS2 are written for each sample to separate files in the <code>results</code> directory. They are named after the corresponding files with extensions used by MACS2.</p>
</div>
<div id="identify-consensus-peaks" class="section level2">
<h2 class="hasAnchor">
<a href="#identify-consensus-peaks" class="anchor"></a>Identify consensus peaks</h2>
<p>The following example shows how one can identify consensus preaks among two peak sets sharing either a minimum absolute overlap and/or minimum relative overlap using the <code>subsetByOverlaps</code> or <code>olRanges</code> functions, respectively. Note, the latter is a custom function imported below by sourcing it.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># source('http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/My_R_Scripts/rangeoverlapper.R')</span>
<span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>  <span class="co">## escolher um dos outputs index</span>
<span class="va">peak_M1A</span> <span class="op">&lt;-</span> <span class="va">outpaths</span><span class="op">[</span><span class="st">"M1A"</span><span class="op">]</span>
<span class="va">peak_M1A</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">peak_M1A</span>, comment <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>, <span class="st">"GRanges"</span><span class="op">)</span>
<span class="va">peak_A1A</span> <span class="op">&lt;-</span> <span class="va">outpaths</span><span class="op">[</span><span class="st">"A1A"</span><span class="op">]</span>
<span class="va">peak_A1A</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">peak_A1A</span>, comment <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>, <span class="st">"GRanges"</span><span class="op">)</span>
<span class="op">(</span><span class="va">myol1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html">subsetByOverlaps</a></span><span class="op">(</span><span class="va">peak_M1A</span>, <span class="va">peak_A1A</span>, minoverlap <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="co"># Returns any overlap</span>
<span class="va">myol2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/olRanges.html">olRanges</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">peak_M1A</span>, subject <span class="op">=</span> <span class="va">peak_A1A</span>, output <span class="op">=</span> <span class="st">"gr"</span><span class="op">)</span>
<span class="co"># Returns any overlap with OL length information</span>
<span class="va">myol2</span><span class="op">[</span><span class="fu">values</span><span class="op">(</span><span class="va">myol2</span><span class="op">)</span><span class="op">[</span><span class="st">"OLpercQ"</span><span class="op">]</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">50</span><span class="op">]</span>
<span class="co"># Returns only query peaks with a minimum overlap of 50%</span></code></pre></div>
</div>
</div>
<div id="annotate-peaks-with-genomic-context" class="section level1">
<h1 class="hasAnchor">
<a href="#annotate-peaks-with-genomic-context" class="anchor"></a>Annotate peaks with genomic context</h1>
<div id="annotation-with-chippeakanno-package" class="section level2">
<h2 class="hasAnchor">
<a href="#annotation-with-chippeakanno-package" class="anchor"></a>Annotation with <code>ChIPpeakAnno</code> package</h2>
<p>The following annotates the identified peaks with genomic context information using the <code>ChIPpeakAnno</code> and <code>ChIPseeker</code> packages, respectively <span class="citation">(Zhu et al. 2010; Yu, Wang, and He 2015)</span>.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ChIPpeakAnno</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures">GenomicFeatures</a></span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/annotate_peaks"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWF</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="st">"targets_macs.txt"</span>, wf_file <span class="op">=</span> <span class="st">"annotate-peaks.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"annotate-peaks.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>

<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html">makeTxDbFromGFF</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/tair10.gff"</span>, format <span class="op">=</span> <span class="st">"gff"</span>, 
    dataSource <span class="op">=</span> <span class="st">"TAIR"</span>, organism <span class="op">=</span> <span class="st">"Arabidopsis thaliana"</span><span class="op">)</span>
<span class="va">ge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html">genes</a></span><span class="op">(</span><span class="va">txdb</span>, columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"tx_name"</span>, <span class="st">"gene_id"</span>, <span class="st">"tx_type"</span><span class="op">)</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">args</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">peaksGR</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs.html">infile1</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, comment <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span>, 
        <span class="st">"GRanges"</span><span class="op">)</span>
    <span class="va">annotatedPeak</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/annotatePeakInBatch.html">annotatePeakInBatch</a></span><span class="op">(</span><span class="va">peaksGR</span>, AnnotationData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html">genes</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span><span class="op">)</span>
    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">annotatedPeak</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu">values</span><span class="op">(</span><span class="va">ge</span><span class="op">[</span><span class="fu">values</span><span class="op">(</span><span class="va">annotatedPeak</span><span class="op">)</span><span class="op">$</span><span class="va">feature</span>, 
        <span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
    <span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">outpaths</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, 
        sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">args</span>, file <span class="op">=</span> <span class="st">"targets_peakanno.txt"</span>, step <span class="op">=</span> <span class="fl">1</span>, 
    new_col <span class="op">=</span> <span class="st">"FileName"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>The peak annotation results are written for each peak set to separate files in the <code>results</code> directory. They are named after the corresponding peak files with extensions specified in the <code>annotate_peaks.param</code> file, here <code>*.peaks.annotated.xls</code>.</p>
</div>
<div id="annotation-with-chipseeker-package" class="section level2">
<h2 class="hasAnchor">
<a href="#annotation-with-chipseeker-package" class="anchor"></a>Annotation with <code>ChIPseeker</code> package</h2>
<p>Same as in previous step but using the <code>ChIPseeker</code> package for annotating the peaks.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ChIPseeker</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">args</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">peakAnno</span> <span class="op">&lt;-</span> <span class="fu">annotatePeak</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs.html">infile1</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, TxDb <span class="op">=</span> <span class="va">txdb</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">peakAnno</span><span class="op">)</span>
    <span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">outpaths</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, 
        sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">args</span>, file <span class="op">=</span> <span class="st">"targets_peakanno.txt"</span>, step <span class="op">=</span> <span class="fl">1</span>, 
    new_col <span class="op">=</span> <span class="st">"FileName"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Summary plots provided by the <code>ChIPseeker</code> package. Here applied only to one sample for demonstration purposes.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">peak</span> <span class="op">&lt;-</span> <span class="fu">readPeakFile</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs.html">infile1</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="fu">covplot</span><span class="op">(</span><span class="va">peak</span>, weightCol <span class="op">=</span> <span class="st">"X.log10.pvalue."</span><span class="op">)</span>
<span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu">peakHeatmap</span><span class="op">(</span><span class="va">outpaths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, TxDb <span class="op">=</span> <span class="va">txdb</span>, upstream <span class="op">=</span> <span class="fl">1000</span>, downstream <span class="op">=</span> <span class="fl">1000</span>, 
    color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span>
<span class="fu">plotAvgProf2</span><span class="op">(</span><span class="va">outpaths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, TxDb <span class="op">=</span> <span class="va">txdb</span>, upstream <span class="op">=</span> <span class="fl">1000</span>, downstream <span class="op">=</span> <span class="fl">1000</span>, 
    xlab <span class="op">=</span> <span class="st">"Genomic Region (5'-&gt;3')"</span>, ylab <span class="op">=</span> <span class="st">"Read Count Frequency"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="count-reads-overlapping-peaks" class="section level1">
<h1 class="hasAnchor">
<a href="#count-reads-overlapping-peaks" class="anchor"></a>Count reads overlapping peaks</h1>
<p>The <code>countRangeset</code> function is a convenience wrapper to perform read counting iteratively over serveral range sets, here peak range sets. Internally, the read counting is performed with the <code>summarizeOverlaps</code> function from the <code>GenomicAlignments</code> package. The resulting count tables are directly saved to files, one for each peak set.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/count_rangesets"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWF</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="st">"targets_macs.txt"</span>, wf_file <span class="op">=</span> <span class="st">"count_rangesets.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"count_rangesets.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>

<span class="co">## Bam Files</span>
<span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE_chip.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/bowtie2/bowtie2-pe"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">args_bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWF</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="va">targets</span>, wf_file <span class="op">=</span> <span class="st">"bowtie2-mapping-pe.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"bowtie2-mapping-pe.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">args_bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args_bam</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">args_bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/output_update.html">output_update</a></span><span class="op">(</span><span class="va">args_bam</span>, dir <span class="op">=</span> <span class="cn">FALSE</span>, replace <span class="op">=</span> <span class="cn">TRUE</span>, 
    extension <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">".sam"</span>, <span class="st">".bam"</span><span class="op">)</span><span class="op">)</span>
<span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args_bam</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>

<span class="va">bfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html">BamFileList</a></span><span class="op">(</span><span class="va">outpaths</span>, yieldSize <span class="op">=</span> <span class="fl">50000</span>, index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="va">countDFnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/countRangeset.html">countRangeset</a></span><span class="op">(</span><span class="va">bfl</span>, <span class="va">args</span>, mode <span class="op">=</span> <span class="st">"Union"</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">args</span>, file <span class="op">=</span> <span class="st">"targets_countDF.txt"</span>, step <span class="op">=</span> <span class="fl">1</span>, 
    new_col <span class="op">=</span> <span class="st">"FileName"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="differential-binding-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-binding-analysis" class="anchor"></a>Differential binding analysis</h1>
<p>The <code>runDiff</code> function performs differential binding analysis in batch mode for several count tables using <code>edgeR</code> or <code>DESeq2</code> <span class="citation">(Robinson, McCarthy, and Smyth 2010; Love, Huber, and Anders 2014)</span>. Internally, it calls the functions <code>run_edgeR</code> and <code>run_DESeq2</code>. It also returns the filtering results and plots from the downstream <code>filterDEGs</code> function using the fold change and FDR cutoffs provided under the <code>dbrfilter</code> argument.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/rundiff"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">args_diff</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWF</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="st">"targets_countDF.txt"</span>, wf_file <span class="op">=</span> <span class="st">"rundiff.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"rundiff.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">args_diff</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args_diff</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>

<span class="va">cmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/readComp.html">readComp</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">args_bam</span>, format <span class="op">=</span> <span class="st">"matrix"</span><span class="op">)</span>
<span class="va">dbrlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runDiff.html">runDiff</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">args_diff</span>, diffFct <span class="op">=</span> <span class="va">run_edgeR</span>, targets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/targets.as.df.html">targets.as.df</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/targets.html">targets</a></span><span class="op">(</span><span class="va">args_bam</span><span class="op">)</span><span class="op">)</span>, 
    cmp <span class="op">=</span> <span class="va">cmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, independent <span class="op">=</span> <span class="cn">TRUE</span>, dbrfilter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>Fold <span class="op">=</span> <span class="fl">2</span>, 
        FDR <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsout.html">writeTargetsout</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">args_diff</span>, file <span class="op">=</span> <span class="st">"targets_rundiff.txt"</span>, 
    step <span class="op">=</span> <span class="fl">1</span>, new_col <span class="op">=</span> <span class="st">"FileName"</span>, new_col_output_index <span class="op">=</span> <span class="fl">1</span>, 
    overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="go-term-enrichment-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#go-term-enrichment-analysis" class="anchor"></a>GO term enrichment analysis</h1>
<p>The following performs GO term enrichment analysis for each annotated peak set.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/annotate_peaks"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWF</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="st">"targets_bam_ref.txt"</span>, wf_file <span class="op">=</span> <span class="st">"annotate-peaks.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"annotate-peaks.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>

<span class="va">args_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWF</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="st">"targets_macs.txt"</span>, wf_file <span class="op">=</span> <span class="st">"annotate-peaks.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"annotate-peaks.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">args_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args_anno</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>
<span class="va">annofiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/subsetWF.html">subsetWF</a></span><span class="op">(</span><span class="va">args_anno</span>, slot <span class="op">=</span> <span class="st">"output"</span>, subset <span class="op">=</span> <span class="fl">1</span>, 
    index <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">gene_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">annofiles</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">annofiles</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span><span class="op">[</span>, 
    <span class="st">"geneId"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/GO/catdb.RData"</span><span class="op">)</span>
<span class="va">BatchResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/GOHyperGAll.html">GOCluster_Report</a></span><span class="op">(</span>catdb <span class="op">=</span> <span class="va">catdb</span>, setlist <span class="op">=</span> <span class="va">gene_ids</span>, 
    method <span class="op">=</span> <span class="st">"all"</span>, id_type <span class="op">=</span> <span class="st">"gene"</span>, CLSZ <span class="op">=</span> <span class="fl">2</span>, cutoff <span class="op">=</span> <span class="fl">0.9</span>, 
    gocats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MF"</span>, <span class="st">"BP"</span>, <span class="st">"CC"</span><span class="op">)</span>, recordSpecGO <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></code></pre></div>
</div>
<div id="motif-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#motif-analysis" class="anchor"></a>Motif analysis</h1>
<div id="parse-dna-sequences-of-peak-regions-from-genome" class="section level2">
<h2 class="hasAnchor">
<a href="#parse-dna-sequences-of-peak-regions-from-genome" class="anchor"></a>Parse DNA sequences of peak regions from genome</h2>
<p>Enrichment analysis of known DNA binding motifs or <em>de novo</em> discovery of novel motifs requires the DNA sequences of the identified peak regions. To parse the corresponding sequences from the reference genome, the <code>getSeq</code> function from the <code>Biostrings</code> package can be used. The following example parses the sequences for each peak set and saves the results to separate FASTA files, one for each peak set. In addition, the sequences in the FASTA files are ranked (sorted) by increasing p-values as expected by some motif discovery tools, such as <code>BCRANK</code>.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings">Biostrings</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">seqLogo</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BCRANK</span><span class="op">)</span>
<span class="va">dir_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl/annotate_peaks"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/loadWorkflow.html">loadWF</a></span><span class="op">(</span>targets <span class="op">=</span> <span class="st">"targets_macs.txt"</span>, wf_file <span class="op">=</span> <span class="st">"annotate-peaks.cwl"</span>, 
    input_file <span class="op">=</span> <span class="st">"annotate-peaks.yml"</span>, dir_path <span class="op">=</span> <span class="va">dir_path</span><span class="op">)</span>
<span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderWF.html">renderWF</a></span><span class="op">(</span><span class="va">args</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, 
    SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span><span class="op">)</span>

<span class="va">rangefiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/sysargs.html">infile1</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">rangefiles</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="va">rangefiles</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, comment <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span>
    <span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">df</span>, <span class="st">"GRanges"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span><span class="op">)</span>, <span class="st">"_"</span>, 
        <span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span>, <span class="st">"-"</span>, <span class="fu"><a href="https://rdrr.io/r/stats/start.html">end</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span><span class="op">)</span>
    <span class="va">peaks</span> <span class="op">&lt;-</span> <span class="va">peaks</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="fu">values</span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span><span class="op">$</span><span class="va">X.log10.pvalue.</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span>
    <span class="va">pseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/getSeq.html">getSeq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/FaFile-class.html">FaFile</a></span><span class="op">(</span><span class="st">"./data/tair10.fasta"</span><span class="op">)</span>, <span class="va">peaks</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">pseq</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">writeXStringSet</a></span><span class="op">(</span><span class="va">pseq</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">rangefiles</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".fasta"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
</div>
<div id="motif-discovery-with-bcrank" class="section level2">
<h2 class="hasAnchor">
<a href="#motif-discovery-with-bcrank" class="anchor"></a>Motif discovery with <code>BCRANK</code>
</h2>
<p>The Bioconductor package <code>BCRANK</code> is one of the many tools available for <em>de novo</em> discovery of DNA binding motifs in peak regions of ChIP-Seq experiments. The given example applies this method on the first peak sample set and plots the sequence logo of the highest ranking motif.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>
<span class="va">BCRANKout</span> <span class="op">&lt;-</span> <span class="fu">bcrank</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">rangefiles</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">".fasta"</span><span class="op">)</span>, restarts <span class="op">=</span> <span class="fl">25</span>, 
    use.P1 <span class="op">=</span> <span class="cn">TRUE</span>, use.P2 <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">toptable</span><span class="op">(</span><span class="va">BCRANKout</span><span class="op">)</span>
<span class="va">topMotif</span> <span class="op">&lt;-</span> <span class="fu">toptable</span><span class="op">(</span><span class="va">BCRANKout</span>, <span class="fl">1</span><span class="op">)</span>
<span class="va">weightMatrix</span> <span class="op">&lt;-</span> <span class="fu">pwm</span><span class="op">(</span><span class="va">topMotif</span>, normalize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">weightMatrixNormalized</span> <span class="op">&lt;-</span> <span class="fu">pwm</span><span class="op">(</span><span class="va">topMotif</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html">pdf</a></span><span class="op">(</span><span class="st">"results/seqlogo.pdf"</span><span class="op">)</span>
<span class="fu">seqLogo</span><span class="op">(</span><span class="va">weightMatrixNormalized</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<img src="seqlogo.png"><div align="center">
Figure 2: One of the motifs identified by <code>BCRANK</code>
</div>
<p></p>
</div>
</div>
<div id="version-information" class="section level1">
<h1 class="hasAnchor">
<a href="#version-information" class="anchor"></a>Version Information</h1>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R Under development (unstable) (2021-02-04 r79940)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /home/dcassol/src/R-devel/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices
## [6] utils     datasets  methods   base     
## 
## other attached packages:
##  [1] systemPipeR_1.25.11         ShortRead_1.49.2           
##  [3] GenomicAlignments_1.27.2    SummarizedExperiment_1.21.3
##  [5] Biobase_2.51.0              MatrixGenerics_1.3.1       
##  [7] matrixStats_0.58.0          BiocParallel_1.25.5        
##  [9] Rsamtools_2.7.2             Biostrings_2.59.2          
## [11] XVector_0.31.1              GenomicRanges_1.43.4       
## [13] GenomeInfoDb_1.27.11        IRanges_2.25.9             
## [15] S4Vectors_0.29.15           BiocGenerics_0.37.2        
## [17] BiocStyle_2.19.2           
## 
## loaded via a namespace (and not attached):
##   [1] colorspace_2.0-0         rjson_0.2.20            
##   [3] hwriter_1.3.2            ellipsis_0.3.1          
##   [5] rprojroot_2.0.2          fs_1.5.0                
##   [7] DT_0.18                  bit64_4.0.5             
##   [9] AnnotationDbi_1.53.1     fansi_0.4.2             
##  [11] codetools_0.2-18         cachem_1.0.4            
##  [13] knitr_1.32               jsonlite_1.7.2          
##  [15] annotate_1.69.2          dbplyr_2.1.1            
##  [17] png_0.1-7                shiny_1.6.0             
##  [19] BiocManager_1.30.12      compiler_4.1.0          
##  [21] httr_1.4.2               backports_1.2.1         
##  [23] assertthat_0.2.1         Matrix_1.3-2            
##  [25] fastmap_1.1.0            limma_3.47.12           
##  [27] later_1.1.0.1            formatR_1.9             
##  [29] htmltools_0.5.1.1        prettyunits_1.1.1       
##  [31] tools_4.1.0              gtable_0.3.0            
##  [33] glue_1.4.2               GenomeInfoDbData_1.2.4  
##  [35] dplyr_1.0.5              rsvg_2.1.1              
##  [37] batchtools_0.9.15        rappdirs_0.3.3          
##  [39] V8_3.4.0                 Rcpp_1.0.6              
##  [41] jquerylib_0.1.3          pkgdown_1.6.1           
##  [43] vctrs_0.3.7              rtracklayer_1.51.5      
##  [45] crosstalk_1.1.1          xfun_0.22               
##  [47] stringr_1.4.0            mime_0.10               
##  [49] lifecycle_1.0.0          restfulr_0.0.13         
##  [51] XML_3.99-0.6             edgeR_3.33.3            
##  [53] scales_1.1.1             zlibbioc_1.37.0         
##  [55] BSgenome_1.59.2          VariantAnnotation_1.37.1
##  [57] promises_1.2.0.1         ragg_1.1.2              
##  [59] hms_1.0.0                RColorBrewer_1.1-2      
##  [61] yaml_2.2.1               curl_4.3                
##  [63] memoise_2.0.0            ggplot2_3.3.3           
##  [65] sass_0.3.1               biomaRt_2.47.7          
##  [67] latticeExtra_0.6-29      stringi_1.5.3           
##  [69] RSQLite_2.2.7            highr_0.9               
##  [71] BiocIO_1.1.2             desc_1.3.0              
##  [73] checkmate_2.0.0          GenomicFeatures_1.43.8  
##  [75] filelock_1.0.2           DOT_0.1                 
##  [77] rlang_0.4.10             pkgconfig_2.0.3         
##  [79] systemfonts_1.0.1        bitops_1.0-6            
##  [81] evaluate_0.14            lattice_0.20-41         
##  [83] purrr_0.3.4              htmlwidgets_1.5.3       
##  [85] bit_4.0.4                tidyselect_1.1.0        
##  [87] magrittr_2.0.1           bookdown_0.22           
##  [89] R6_2.5.0                 generics_0.1.0          
##  [91] base64url_1.4            DelayedArray_0.17.10    
##  [93] DBI_1.1.1                pillar_1.6.0            
##  [95] withr_2.4.2              KEGGREST_1.31.1         
##  [97] RCurl_1.98-1.3           tibble_3.1.1            
##  [99] crayon_1.4.1             utf8_1.2.1              
## [101] BiocFileCache_1.99.6     rmarkdown_2.7           
## [103] jpeg_0.1-8.1             progress_1.2.2          
## [105] locfit_1.5-9.4           grid_4.1.0              
## [107] data.table_1.14.0        blob_1.2.1              
## [109] digest_0.6.27            xtable_1.8-4            
## [111] httpuv_1.5.5             brew_1.0-6              
## [113] textshaping_0.3.3        munsell_0.5.0           
## [115] bslib_0.2.4</code></pre>
</div>
<div id="funding" class="section level1">
<h1 class="hasAnchor">
<a href="#funding" class="anchor"></a>Funding</h1>
<p>This project was supported by funds from the National Institutes of Health (NIH) and the National Science Foundation (NSF).</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Buenrostro2013-ap" class="csl-entry">
Buenrostro, Jason D, Paul G Giresi, Lisa C Zaba, Howard Y Chang, and William J Greenleaf. 2013. <span>“Transposition of Native Chromatin for Fast and Sensitive Epigenomic Profiling of Open Chromatin, <span class="nocase">DNA-binding</span> Proteins and Nucleosome Position.”</span> <em>Nat. Methods</em> 10 (12): 1213–18.
</div>
<div id="ref-H_Backman2016-bt" class="csl-entry">
H Backman, Tyler W, and Thomas Girke. 2016. <span>“<span class="nocase">systemPipeR: NGS workflow and report generation environment</span>.”</span> <em>BMC Bioinformatics</em> 17 (1): 388. <a href="https://doi.org/10.1186/s12859-016-1241-0">https://doi.org/10.1186/s12859-016-1241-0</a>.
</div>
<div id="ref-Love2014-sh" class="csl-entry">
Love, Michael, Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for <span class="nocase">RNA-seq</span> Data with <span>DESeq2</span>.”</span> <em>Genome Biol.</em> 15 (12): 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-Robinson2010-uk" class="csl-entry">
Robinson, M D, D J McCarthy, and G K Smyth. 2010. <span>“edgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.”</span> <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.
</div>
<div id="ref-Yu2015-xu" class="csl-entry">
Yu, Guangchuang, Li-Gen Wang, and Qing-Yu He. 2015. <span>“<span>ChIPseeker</span>: An <span>R/Bioconductor</span> Package for <span>ChIP</span> Peak Annotation, Comparison and Visualization.”</span> <em>Bioinformatics</em> 31 (14): 2382–83. <a href="https://doi.org/10.1093/bioinformatics/btv145">https://doi.org/10.1093/bioinformatics/btv145</a>.
</div>
<div id="ref-Zhang2008-pc" class="csl-entry">
Zhang, Y, T Liu, C A Meyer, J Eeckhoute, D S Johnson, B E Bernstein, C Nussbaum, et al. 2008. <span>“Model-Based Analysis of <span>ChIP-Seq</span> (<span>MACS</span>).”</span> <em>Genome Biol.</em> 9 (9). <a href="https://doi.org/10.1186/gb-2008-9-9-r137">https://doi.org/10.1186/gb-2008-9-9-r137</a>.
</div>
<div id="ref-Zhu2010-zo" class="csl-entry">
Zhu, Lihua J, Claude Gazin, Nathan D Lawson, Hervé Pagès, Simon M Lin, David S Lapointe, and Michael R Green. 2010. <span>“<span>ChIPpeakAnno</span>: A Bioconductor Package to Annotate <span class="nocase">ChIP-seq</span> and <span class="nocase">ChIP-chip</span> Data.”</span> <em>BMC Bioinformatics</em> 11: 237. <a href="https://doi.org/10.1186/1471-2105-11-237">https://doi.org/10.1186/1471-2105-11-237</a>.
</div>
<div id="ref-Ziemann2016-mp" class="csl-entry">
Ziemann, Mark, Antony Kaspi, and Assam El-Osta. 2016. <span>“Evaluation of <span class="nocase">microRNA</span> Alignment Techniques.”</span> <em>RNA</em> 22 (8): 1120–38.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://www.dcassol.com/">Daniela Cassol</a>, <a href="http://girke.bioinformatics.ucr.edu/">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
